packages/core/core/generate-object/stream-object.ts

```diff
@@ -361,24 +359,27 @@ Response headers.
     let error: unknown | undefined;
 
     // pipe chunks through a transformation stream that extracts metadata:
-    let accumulatedText = '';
     let latestObject: DeepPartial<T> | undefined = undefined;
 
+    const streamingParser = new StreamingParser({ stream: true });
+
     this.originalStream = stream.pipeThrough(
       new TransformStream<string | ObjectStreamInputPart, ObjectStreamPart<T>>({
         async transform(chunk, controller): Promise<void> {
           // process partial text chunks
           if (typeof chunk === 'string') {
-            accumulatedText += chunk;
-
-            const currentObject = parsePartialJson(
-              accumulatedText,
-            ) as DeepPartial<T>;
-
-            if (!isDeepEqualData(latestObject, currentObject)) {
-              latestObject = currentObject;
-
-              controller.enqueue({ type: 'object', object: currentObject });
+            let { events } = streamingParser.parseIncremental(chunk);
+
+            // filter for events that change values
+            // @ts-expect-error TODO: fix type
+            events = events.filter(event => event.kind === 'partial' || (event.kind === 'complete' && typeof event.value !== 'object'))
```

Comment:
The `ParseEvent<T>` is richly typed, though TypeScript's narrowing is not quite powerful enough without a bit of help to provide the right type, though perhaps the current type could be updated to help TSC perform that narrowing.

The `ts-patterns` library does do this, and shows what should be possible, as the type of `event.value` here is narrowed typed:

https://github.com/AaronFriel/fn-stream/blob/a208bf2f5dd686cf4e6e4f5c78d8e45e445daa7f/packages/fn-stream/examples/openai-reducer.spec.ts#L108-L128